bolt-modules/boltlib/lib/puppet/datatypes/applyresult.rb

```diff
@@ -33,6 +41,7 @@
       'target' => Target
     },
     functions => {
+      catalog => Callable[[], Optional[Hash[String[1], Data]]],
```

Comment:
Making this a function is a little weird, because it means the `ApplyResult` object can't round trip through JSON. Of course it couldn't anyway because the catalog is sensitive. But it does seem to break one of the rules of pcore data types, which is that functions should only return data that can be derived from the attributes. In this case, we effectively sideload some additional data into the report that is not one of the attributes. I'm not sure what to make of that though.

Separately, I wonder if the specificity of this type declaration has a performance impact. When the function is called, puppet should verify that the result conforms to the type. In this case, that means verifying that every hash value is Data, which means traversing the entire catalog down to the leaves. Since the catalog is quite a large object, that could be significant. Of course, it only affects users who actually _call_ the `catalog()` function, so it's not a critical problem. It would be interesting to test though. If it's a problem, just declaring this as a `Hash` ought to be sufficient.

Reply:
@nicklewis Wouldn't that also apply to `error`? Or am I misunderstanding what is considered an attribute in this case?

I agree that just declaring the catalog a `Hash` would be sufficient in this case.

Reply:
Hmm yes that would also apply to `error`. We do return `error` from `_pcore_init_hash`, so I wonder why that doesn't cause a problem when we try to serialize and deserialize an ApplyResult. For instance, if you run multiple `apply()` blocks in a single plan, such that the result of one is available within the next one.